src/paste-markdown-html.ts

```diff
@@ -20,65 +18,80 @@ function onPaste(event: ClipboardEvent) {
-  elements: HTMLCollectionOf<HTMLElement>,
-  text: string,
-  transformer: MarkdownTransformer,
-  ...args: string[]
-): string {
-  const markdownParts = []
-  for (const element of elements) {
-    const textContent = element.textContent || ''
-    const {part, index} = trimAfter(text, textContent)
-    if (index >= 0) {
-      markdownParts.push(part.replace(textContent, transformer(element, args)))
-      text = text.slice(index)
+function convertToMarkdown(plaintext: string, walker: TreeWalker): string {
+  let currentNode = walker.firstChild() as HTMLAnchorElement | HTMLElement | null
+  let markdown = plaintext
+  let markdownIgnoreBeforeIndex = 0
+  let index = 0
+  const NODE_LIMIT = 100000
+
+  // Walk through the DOM tree
+  while (currentNode && index < NODE_LIMIT) {
+    index++
+    const text = isLink(currentNode)
+      ? (currentNode as HTMLAnchorElement).textContent || ''
+      : (currentNode.firstChild as Text)?.wholeText || ''
+
+    // No need to transform whitespace
+    if (isEmptyString(text)) {
+      currentNode = walker.nextNode() as HTMLAnchorElement | HTMLElement | null
+      continue
+    }
+
+    // Find the index where "text" is found in "markdown" _after_ "markdownIgnoreBeforeIndex"
+    const markdownFoundIndex = markdown.indexOf(text, markdownIgnoreBeforeIndex)
+
+    if (markdownFoundIndex >= 0) {
+      if (isLink(currentNode)) {
+        const markdownLink = linkify(currentNode as HTMLAnchorElement)
+        // Transform 'example link plus more text' into 'example [link](example link) plus more text'
+        // Method: 'example [link](example link) plus more text' = 'example ' + '[link](example link)' + ' plus more text'
+        markdown =
+          markdown.slice(0, markdownFoundIndex) + markdownLink + markdown.slice(markdownFoundIndex + text.length)
+        markdownIgnoreBeforeIndex = markdownFoundIndex + markdownLink.length
+      } else {
+        markdownIgnoreBeforeIndex = markdownFoundIndex + text.length
+      }
     }
+
+    currentNode = walker.nextNode() as HTMLAnchorElement | HTMLElement | null
   }
-  markdownParts.push(text)
-  return markdownParts.join('')
-}
 
-// Trim text at index of last character of the first occurrence of "search" and
-// return a new string with the substring until the index
-//  Example: trimAfter('Hello world', 'world') => {part: 'Hello world', index: 11}
-//  Example: trimAfter('Hello world', 'bananas') => {part: '', index: -1}
-function trimAfter(text: string, search = ''): {part: string; index: number} {
-  let index = text.indexOf(search)
-  if (index === -1) return {part: '', index}
+  // Unless we hit the node limit, we should have processed all nodes
+  return index === NODE_LIMIT ? plaintext : markdown
+}
 
-  index += search.length
+function isEmptyString(text: string): boolean {
+  return !text || text?.trim().length === 0
+}
 
-  return {
-    part: text.substring(0, index),
-    index
-  }
+function isLink(node: HTMLElement): boolean {
```

Comment:
If you change this to 
```suggestion
function isLink(node: HTMLElement): node is HTMLAnchorElement {
```

Then it'll refine the type that TypeScript can resolve, which saves a bunch of `as HTMLAnchorElement` casts you have.